### PR TITLE
integrating utf8 validation

### PR DESCRIPTION
**Pull request by weissi**
together with https://github.com/milseman/UTF8String/pull/10

### CODE REVIEW

**Comment by milseman** [comment]
Yes, I think that's along the right lines. We might want to have a helper function that does the loop+correct. If it's easy for the validator to remember certain properties (such as ASCII-ness), that'd be nice to report back, otherwise it's still worth doing a second pass dedicated to determining ASCII-ness.

**Comment by weissi** [comment]
@milseman agreed! I need to leave for tonight unfortunately but tomorrow I'll prepare a patch to extract the loop+correct. Straight after that I propose to make a new PR to remember the ASCII-ness etc in the validator. Does that sound okay or would you prefer me to do that in this PR?

**Comment by milseman** [comment]
Sounds good to me.

File: stdlib/public/core/StringUTF8Validation.swift [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+@usableFromInline
+func utf8Repair(_ input: UnsafeBufferPointer<UInt8>, firstKnownBrokenRange: Range<Int>) -> String {
+  var x = String()
+  let replacementCharacterCount = utf8ReplacementCharacter.count
+  x.reserveCapacity(input.count + 5 * replacementCharacterCount) // just some extra space
+
+  var lastGoodIndex = 0
+  var brokenRange: Range<Int> = firstKnownBrokenRange
+  var goodRange: Range<Int> = lastGoodIndex ..< brokenRange.startIndex
+  loop: repeat {
```
Comment (weissi) [inline stdlib/public/core/StringUTF8Validation.swift]: @milseman sorry, this stuff is pretty ugly but should be much faster than the old one as it no longer uses an extra array, so removes an allocation and a copy.

Comment (weissi) [reply stdlib/public/core/StringUTF8Validation.swift]: @milseman if I clean this up a bit (renaming `x`, and giving the various ranges better names), is this roughly what you thought of?

**Comment by weissi** [comment]
@milseman thanks very much for your comments, updated

File: stdlib/public/core/StringUTF8Validation.swift [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+      if _isASCII(cu) { lastValidIndex &+= 1; continue }
+      isASCII = false
+      if _slowPath(!_isUTF8MultiByteLeading(cu)) {
+        throw UTF8ValidationError()
+      }
+      // TODO: perhaps a custom branching pattern
+      switch cu {
+      case 0xC2...0xDF:
+        try guaranteeContinuation()
+        lastValidIndex &+= 2
```
Comment (weissi) [inline stdlib/public/core/StringUTF8Validation.swift]: note for @milseman : this is new in this version: no longer checking overflow here, therefore made `lastValidIndex` a `UInt`, so overflow check moves down (in only one place now, where we make that an `Int`)

Comment (weissi) [reply stdlib/public/core/StringUTF8Validation.swift]: Ok cool, just didn’t want to introduce extra restrictions but more than happy to make it `Int`. Will do tomorrow when I’m back at a computer

File: stdlib/public/core/StringUTF8Validation.swift [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+private func _isNotOverlong_ED(_ x: UInt8) -> Bool {
+  return _isContinuation(x) && x <= 0x9F
+}
+
+private func _isASCII_cmp(_ x: UInt8) -> Bool {
+  return x <= 0x7F
+}
+
+@usableFromInline
```
[the remark this answers is not in the document]
Comment (weissi) [reply stdlib/public/core/StringUTF8Validation.swift]: Ok, I great, I really like explicit `internal`. Sad that Swift doesn’t default to private

**Comment by weissi** [comment]
@milseman thanks again! Updated. I did change the `lastValidIndex` to an `Int` now. Do you want me to add any more `_sanityChecks` for that?

File: stdlib/public/core/StringUTF8Validation.swift [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    switch validateUTF8(remainder) {
+    case .success:
+      result._guts.appendInPlace(remainder, isASCII: false)
+      return result
+    case .error(let newBrokenRangeRebased):
+      brokenRange = (newBrokenRangeRebased.startIndex + remainderRange.startIndex) ..< (newBrokenRangeRebased.endIndex + remainderRange.startIndex)
+      goodRange = remainderRange.startIndex ..< brokenRange.startIndex
+    }
+  } while true
+}
```
Comment (milseman) [inline stdlib/public/core/StringUTF8Validation.swift]: I think this method can be simplified. You don't need to track so many ranges if the buffer pointer is the loop-variant and gets rebased each iteration.

Something like (pseudo-y):

```
var result = _StringGuts()
var input = input
var brokenRange = firstKnownBrokenRange
while true {
  result._appendInPlace(UBP(rebasing: input[..<brokenRange.lowerBound]))
  U.S._replacementCharacter.withUTF8CodeUnits { result._appendInPlace($0) }
  input = UBP(rebasing: input[brokenRange.upperBound...])
  switch validateUTF8(input) { // Well, assuming validateUTF8 returns success for empty range, we should test this!
    case .success:
      result._appendInPlace(input, isASCII: false)
      return String(result)
    case .error(let newBrokenRange):
      brokenRange = newBrokenRange
  }
}
return String(result)
```

Comment (milseman) [reply stdlib/public/core/StringUTF8Validation.swift]: Err, with the appropriate reserveCapacity calls in there, of course...

Comment (weissi) [reply stdlib/public/core/StringUTF8Validation.swift]: - we have an empty  range check
- I came up with something close to that. `brokenRange = newBrokenRange` wouldn't work as we need to "un-rebase" it first 😤 (such a shame that the UBP indices aren't the pointer value which causes this whole rebase-mess)

Comment (milseman) [reply stdlib/public/core/StringUTF8Validation.swift]: That's why we rebase input, so that brokenRange is always relative to the input range.

**Comment by weissi** [comment]
@milseman testing notifications & updated the PR

**Pull request merged**